### PR TITLE
Avoid deployment distribution on reprocessing

### LINKED ISSUE #3124: Deployments are re-distributed during reprocessing
**Description**

During reprocessing of partition 1, the deployments are pushed to other partitions again. While it doesn't lead to any incorrect state, there is room for improvement.

```
11:59:09.987 [partition-1-processor] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 1095222022840 
11:59:10.117 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813685250 
11:59:10.689 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813688855 
11:59:25.087 [partition-1-processor] WARN  io.zeebe.broker.workflow.repository - Failed to receive deployment response for partitions [3] (topic 'deployment-response-2251799813685250'). Retrying 
11:59:25.827 [partition-1-processor] WARN  io.zeebe.broker.workflow.repository - Failed to receive deployment response for partitions [3] (topic 'deployment-response-2251799813688855'). Retrying 
11:59:45.177 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813685250 
11:59:45.178 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Deployment 2251799813685250 pushed to all partitions successfully. 
11:59:45.178 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813688855 
11:59:45.178 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Deployment 2251799813688855 pushed to all partitions successfully. 
11:59:45.178 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813685250 
11:59:45.179 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813688855 
11:59:52.288 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813685250 
11:59:53.033 [partition-1-processor] DEBUG io.zeebe.broker.workflow.repository - Receiving deployment response on topic deployment-response-2251799813688855 
12:00:50.550 [partition-1-processor] INFO  io.zeebe.processor - Processor finished reprocessing at event position 1095222022840
```

### PR DESCRIPTION
**Pull request by saig0**
## Description

* don't distribute the deployment always on reprocessing because it may be already reprocessed
* instead, we wait until the reprocessing is done and distribute the pending deployments again (current behavior)

## Related issues

closes #3124 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/processing/deployment/distribute/DeploymentDistributeProcessor.java [hunk 1]
```diff
@@ -75,7 +75,13 @@ public void processRecord(
     final DirectBuffer bufferView = new UnsafeBuffer();
     bufferView.wrap(buffer, 0, deploymentEvent.getLength());
 
-    distributeDeployment(key, position, bufferView, streamWriter);
+    // don't distribute the deployment on reprocessing because it may be already distributed
+    // instead, distribute the pending deployments after the reprocessing
```
Comment (Zelldon) [inline engine/src/main/java/io/zeebe/engine/processing/deployment/distribute/DeploymentDistributeProcessor.java:79]: This is not really the reason right. The reason is that we will write concurrently/async a `DISTRIBUTED` on reprocessing which totally breaks the detection.

Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/processing/deployment/distribute/DeploymentDistributeProcessor.java:79]: Maybe we add a link to the github issue

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java [hunk 2]
```diff
@@ -83,8 +83,7 @@ public void setup() {
   @Test
   public void shouldIgnoreDeploymentDistribution() {
     // given
-    // force that a deployment DISTRIBUTED event is written on reprocessing which was not written on
-    // the log stream before
+    // the deployment can be distributed on reprocessing
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java:86]: Not sure whether this test makes still sense? But I thought maybe we can have a different test where we verify that we not call distribute on reprocessing. I have new logic where I can await reprocessing, which means we know whether before distribute was called. What do you think?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @saig0 

Small comments see below

File: engine/src/main/java/io/zeebe/engine/processing/deployment/distribute/DeploymentDistributeProcessor.java (hunk 1 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processing/deployment/distribute/DeploymentDistributeProcessor.java:79]: It is for both reasons :sweat_smile:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java (hunk 2 above)
Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java:86]: I think the test still makes sense for the detection. However, I can add another that `DeploymentDistributor` is not called on reprocessing. Please point me to the example :)

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java:86]: https://github.com/zeebe-io/zeebe/pull/5761/files#diff-b31c3c84b754243e651147040006e4acd695b0315a1f9832f63b6e20b0a3f7f3R199-R204

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java:86]: I think we wait until this PR is merged than you can rebase

Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReprocessingIssueDetectionMultiplePartitionTest.java]: Ok Chris :sweat_smile: I rewrote the test to verify that the deployment distributor is not invoked on reprocessing. Please have a look :)

File: engine/src/test/java/io/zeebe/engine/processing/deployment/DeploymentReprocessingTest.java [hunk 3]
```diff
@@ -33,82 +30,57 @@
...
-                .limit(r -> r.getPosition() > jobCreated.getPosition())
-                .withValueType(ValueType.DEPLOYMENT))
-        .extracting(Record::getIntent)
-        .doesNotContain(DeploymentIntent.DISTRIBUTED);
+    RecordingExporter.deploymentRecords(DeploymentIntent.DISTRIBUTED).await();
 
     engine.stop();
+
+    // the deployment is distributed once on processing
+    verify(deploymentDistributorMock, times(DEPLOYMENT_DISTRIBUTION_ON_PROCESSING))
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/processing/deployment/DeploymentReprocessingTest.java:66]: We could reset the count after verification. What do you think?

Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/deployment/DeploymentReprocessingTest.java]: Good point. I applied your hint. Please have a look :eyes:

**Review by ChrisKujawa — APPROVED** [review]
Awesome thanks :+1:

**Pull request merged**